File: linux/src/meeting_recorder/daemon/dbus_service.py
"""
The D-Bus Engine service: the daemon<->window boundary.

Exposes ``io.github.jmarceno.Gravaai.Engine`` on the session bus.
Method calls from the window process are routed to the in-daemon ``Engine``;
state changes are pushed back as ``SnapshotChanged`` signals (plus ``Error`` /
``Output`` / ``OpenUseExisting`` / ``PresentWindow``). Also owns the window
child via ``WindowSupervisor`` (spawn on ``OpenWindow``, present the existing
one otherwise).

This module is GTK-free (``Gio``/``GLib`` only) and is not unit-tested — it
needs a real session bus. The spawn/present decision it delegates to
``WindowSupervisor`` is unit-tested separately; the pure snapshot payload it
carries is covered by ``core/wire`` tests.
"""

from __future__ import annotations

import logging
import sys

from gi.repository import Gio, GLib

from ..config.defaults import APP_ID
from .engine import Engine
from .window_supervisor import WindowSupervisor

logger = logging.getLogger(__name__)

ENGINE_NAME = APP_ID
ENGINE_PATH = "/" + APP_ID.replace(".", "/")
ENGINE_IFACE = APP_ID + ".Engine"

_ENGINE_XML = f"""
<node>
  <interface name="{ENGINE_IFACE}">
    <method name="StartRecording"><arg name="mode" type="s" direction="in"/></method>
    <method name="SetTitle"><arg name="title" type="s" direction="in"/></method>
    <method name="Pause"/>
    <method name="Resume"/>
    <method name="Stop"/>
    <method name="CancelCountdown"/>
    <method name="CancelSave"/>
    <method name="Cancel"/>
    <method name="ImportExisting">
      <arg name="audio" type="s" direction="in"/>
      <arg name="transcript" type="s" direction="in"/>
      <arg name="notes" type="s" direction="in"/>
      <arg name="label" type="s" direction="in"/>
    </method>
    <method name="SummarizeMeeting">
      <arg name="audio" type="s" direction="in"/>
      <arg name="transcript" type="s" direction="in"/>
      <arg name="notes" type="s" direction="in"/>
      <arg name="label" type="s" direction="in"/>
      <arg name="error" type="s" direction="out"/>
    </method>
    <method name="CancelJob"><arg name="id" type="i" direction="in"/></method>
    <method name="RetryJob"><arg name="id" type="i" direction="in"/></method>
    <method name="DismissJob"><arg name="id" type="i" direction="in"/></method>
    <method name="JobFolder">
      <arg name="id" type="i" direction="in"/>
      <arg name="path" type="s" direction="out"/>
    </method>
    <method name="OutputFolder"><arg name="path" type="s" direction="out"/></method>
    <method name="ReloadConfig"/>
    <method name="OpenWindow"/>
    <method name="GetSnapshot"><arg name="json" type="s" direction="out"/></method>
    <method name="StartInstall"><arg name="spec" type="s" direction="in"/></method>
    <method name="GetInstalls"><arg name="json" type="s" direction="out"/></method>
    <method name="Quit"/>
    <signal name="SnapshotChanged"><arg name="json" type="s"/></signal>
    <signal name="Error"><arg name="msg" type="s"/></signal>
    <signal name="Output"><arg name="text" type="s"/></signal>
    <signal name="OpenUseExisting"/>
    <signal name="PresentWindow"/>
    <signal name="InstallProgress"><arg name="key" type="s"/><arg name="text" type="s"/></signal>
    <signal name="InstallFinished">
      <arg name="key" type="s"/><arg name="ok" type="b"/><arg name="message" type="s"/>
    </signal>
  </interface>
</node>
"""


class EngineService:
    """Publishes the Engine on D-Bus and supervises the window child."""

    def __init__(
        self,
        engine: Engine,
        on_quit,
        on_reload_config,
        install_manager=None,
    ) -> None:
        self._engine = engine
        self._on_quit = on_quit
        self._on_reload_config = on_reload_config
        self._install_manager = install_manager
        self._conn: Gio.DBusConnection | None = None
        self._reg_id = 0
        self._owner_id = 0
        self._window_proc: Gio.Subprocess | None = None
        self._supervisor = WindowSupervisor(
            spawn_fn=self._spawn_window, present_fn=self._emit_present
        )

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self) -> None:
        self._conn = Gio.bus_get_sync(Gio.BusType.SESSION, None)
        node = Gio.DBusNodeInfo.new_for_xml(_ENGINE_XML)
        self._reg_id = self._conn.register_object(
            ENGINE_PATH, node.interfaces[0], self._on_method, None, None
        )
        self._owner_id = Gio.bus_own_name_on_connection(
            self._conn,
            ENGINE_NAME,
            Gio.BusNameOwnerFlags.NONE,
            None,
            self._on_name_lost,
        )
        logger.info("Engine service registered at %s", ENGINE_PATH)

    def _on_name_lost(self, _conn, _name) -> None:
        # Another daemon owns the name (or the bus went away). A second daemon
        # must not fight for it — log and let this one exit cleanly.
        logger.warning(
            "Lost/failed to acquire bus name %s — another daemon is running", ENGINE_NAME
        )

    # ------------------------------------------------------------------
    # Signals pushed to the window
    # ------------------------------------------------------------------

    def emit_snapshot(self) -> None:
        self._emit("SnapshotChanged", GLib.Variant("(s)", (self._engine.snapshot_json(),)))

    def emit_error(self, msg: str) -> None:
        self._emit("Error", GLib.Variant("(s)", (msg,)))

    def emit_output(self, text: str) -> None:
        self._emit("Output", GLib.Variant("(s)", (text,)))

    def emit_install_progress(self, key: str, text: str) -> None:
        self._emit("InstallProgress", GLib.Variant("(ss)", (key, text)))

    def emit_install_finished(self, key: str, ok: bool, message: str) -> None:
        self._emit("InstallFinished", GLib.Variant("(sbs)", (key, ok, message)))

    def _emit_present(self) -> None:
        self._emit("PresentWindow", None)

    def _emit(self, name: str, body) -> None:
        if self._conn is None:
            return
        try:
            self._conn.emit_signal(None, ENGINE_PATH, ENGINE_IFACE, name, body)
        except GLib.Error as exc:
            logger.debug("Failed to emit %s: %s", name, exc)

    # ------------------------------------------------------------------
    # Window supervision
    # ------------------------------------------------------------------

    def open_window(self) -> None:
        self._supervisor.open()

    def _spawn_window(self) -> None:
        # fork+exec a fresh interpreter — NEVER a bare fork (the daemon has
        # threads, D-Bus connections and a live ffmpeg child that must not be
        # inherited). Gio.Subprocess does fork+exec.
        try:
            proc = Gio.Subprocess.new(
                [sys.executable, "-m", "meeting_recorder", "--window"],
                Gio.SubprocessFlags.NONE,
            )
            self._window_proc = proc
            proc.wait_async(None, self._on_window_exit)
        except GLib.Error as exc:
            logger.error("Failed to spawn window: %s", exc)
            self._window_proc = None
            self._supervisor.on_child_exit()

    def _on_window_exit(self, proc, result) -> None:
        try:
            proc.wait_finish(result)  # reap — no zombies
        except GLib.Error:
            pass
        if proc is self._window_proc:
            self._window_proc = None
        self._supervisor.on_child_exit()

    def shutdown_window(self) -> None:
        """Terminate the window child, if any, before the daemon exits.

        A kept-in-memory (hidden) window would otherwise be orphaned and linger
        after the daemon quits. The window also self-exits when it sees the bus
        name vanish; this makes the cleanup immediate on a clean quit.
        """
        proc = self._window_proc
        if proc is None:
            return
        self._window_proc = None
        try:
            proc.force_exit()
        except GLib.Error as exc:
            logger.debug("Failed to terminate window child: %s", exc)

    # ------------------------------------------------------------------
    # Method dispatch (window -> engine)
    # ------------------------------------------------------------------

    def _on_method(self, _conn, _sender, _path, _iface, method, params, invocation):
        try:
            self._dispatch(method, params, invocation)
        except Exception:
            logger.exception("Engine method %s failed", method)
            invocation.return_value(None)

    def _dispatch(self, method, params, invocation):
        eng = self._engine
        if method == "StartRecording":
            (mode,) = params.unpack()
            eng.start_recording(mode)
            invocation.return_value(None)
        elif method == "SetTitle":
            (title,) = params.unpack()
            eng.set_title(title)
            invocation.return_value(None)
        elif method == "Pause":
            eng.pause()
            invocation.return_value(None)
        elif method == "Resume":
            eng.resume()
            invocation.return_value(None)
        elif method == "Stop":
            eng.stop()
            invocation.return_value(None)
        elif method == "CancelCountdown":
            eng.cancel_countdown()
            invocation.return_value(None)
        elif method == "CancelSave":
            eng.cancel_and_save()
            invocation.return_value(None)
        elif method == "Cancel":
            eng.cancel_and_discard()
            invocation.return_value(None)
        elif method == "ImportExisting":
            audio, transcript, notes, label = params.unpack()
            eng.import_existing(audio, transcript, notes, label)
            invocation.return_value(None)
        elif method == "SummarizeMeeting":
            audio, transcript, notes, label = params.unpack()
            err = eng.summarize_meeting(audio, transcript, notes, label) or ""
            invocation.return_value(GLib.Variant("(s)", (err,)))
        elif method == "CancelJob":
            (jid,) = params.unpack()
            eng.cancel_job(int(jid))
            invocation.return_value(None)
        elif method == "RetryJob":
            (jid,) = params.unpack()
            eng.retry_job(int(jid))
            invocation.return_value(None)
        elif method == "DismissJob":
            (jid,) = params.unpack()
            eng.dismiss_job(int(jid))
            invocation.return_value(None)
        elif method == "JobFolder":
            (jid,) = params.unpack()
            invocation.return_value(GLib.Variant("(s)", (eng.job_folder(int(jid)) or "",)))
        elif method == "OutputFolder":
            invocation.return_value(GLib.Variant("(s)", (eng.output_folder(),)))
        elif method == "ReloadConfig":
            self._on_reload_config()
            invocation.return_value(None)
        elif method == "OpenWindow":
            self.open_window()
            invocation.return_value(None)
        elif method == "GetSnapshot":
            invocation.return_value(GLib.Variant("(s)", (eng.snapshot_json(),)))
        elif method == "StartInstall":
            (spec,) = params.unpack()
            if self._install_manager is not None:
                try:
                    self._install_manager.start(spec)
                except ValueError as exc:
                    logger.warning("Rejected install request: %s", exc)
            invocation.return_value(None)
        elif method == "GetInstalls":
            running = self._install_manager.running_json() if self._install_manager else "[]"
            invocation.return_value(GLib.Variant("(s)", (running,)))
        elif method == "Quit":
            invocation.return_value(None)
            self._on_quit()
        else:
            invocation.return_value(None)

File: linux/src/daemon/dbus_service.rs
//! The D-Bus Engine service: the daemon↔window boundary.
//!
//! Exposes `io.github.jmarceno.GravaAi.Engine` on the session bus. Method
//! calls from the window process are routed to the in-daemon [`Engine`];
//! state changes are pushed back as `SnapshotChanged` signals (plus `Error` /
//! `Output` / `PresentWindow` / install signals) by the daemon loop, which
//! observes engine mutations through the engine hooks. Also owns the window
//! child via [`WindowSupervisor`] (spawn on `OpenWindow`, present the
//! existing one otherwise).

use std::sync::{Arc, Mutex};

use crate::audio::recorder::Recorder;
use crate::config::defaults::APP_ID;
use crate::daemon::engine::Engine;
use crate::daemon::install_manager::InstallManager;
use crate::daemon::window_supervisor::WindowSupervisor;

pub const ENGINE_NAME: &str = APP_ID;
pub const ENGINE_PATH: &str = "/io/github/jmarceno/GravaAi";

/// Shared daemon state behind the D-Bus interface.
pub struct ServiceCtx {
    pub engine: tokio::sync::Mutex<Engine<Recorder>>,
    pub installs: tokio::sync::Mutex<InstallManager>,
    pub supervisor: Mutex<WindowSupervisor>,
    pub on_quit: Box<dyn Fn() + Send + Sync>,
    pub on_reload: Box<dyn Fn() + Send + Sync>,
}

pub struct EngineIface {
    ctx: Arc<ServiceCtx>,
}

impl EngineIface {
    pub fn new(ctx: Arc<ServiceCtx>) -> Self {
        Self { ctx }
    }
}

#[zbus::interface(name = "io.github.jmarceno.GravaAi.Engine")]
impl EngineIface {
    async fn start_recording(&self, mode: String) {
        let mut engine = self.ctx.engine.lock().await;
        engine.reload_config();
        engine.start_recording(&mode);
    }

    async fn set_title(&self, title: String) {
        self.ctx.engine.lock().await.set_title(&title);
    }

    async fn pause(&self) {
        self.ctx.engine.lock().await.pause();
    }

    async fn resume(&self) {
        self.ctx.engine.lock().await.resume();
    }

    async fn stop(&self) {
        let mut engine = self.ctx.engine.lock().await;
        engine.reload_config();
        engine.stop();
    }

    async fn cancel_countdown(&self) {
        self.ctx.engine.lock().await.cancel_countdown();
    }

    #[zbus(name = "CancelSave")]
    async fn cancel_save(&self) {
        self.ctx.engine.lock().await.cancel_and_save();
    }

    #[zbus(name = "Cancel")]
    async fn cancel(&self) {
        self.ctx.engine.lock().await.cancel_and_discard();
    }

    async fn import_existing(
        &self,
        audio: String,
        transcript: String,
        notes: String,
        label: String,
    ) {
        self.ctx
            .engine
            .lock()
            .await
            .import_existing(&audio, &transcript, &notes, &label);
    }

    async fn summarize_meeting(
        &self,
        audio: String,
        transcript: String,
        notes: String,
        label: String,
    ) -> String {
        self.ctx
            .engine
            .lock()
            .await
            .summarize_meeting(&audio, &transcript, &notes, &label)
            .unwrap_or_default()
    }

    async fn cancel_job(&self, id: i32) {
        self.ctx.engine.lock().await.cancel_job(id as i64);
    }

    async fn retry_job(&self, id: i32) {
        self.ctx.engine.lock().await.retry_job(id as i64);
    }

    async fn dismiss_job(&self, id: i32) {
        self.ctx.engine.lock().await.dismiss_job(id as i64);
    }

    async fn job_folder(&self, id: i32) -> String {
        self.ctx
            .engine
            .lock()
            .await
            .job_folder(id as i64)
            .unwrap_or_default()
    }

    async fn output_folder(&self) -> String {
        self.ctx.engine.lock().await.output_folder()
    }

    async fn reload_config(&self) {
        self.ctx.engine.lock().await.reload_config();
        (self.ctx.on_reload)();
    }

    async fn open_window(&self) {
        // Spawn-vs-present decision; the actual spawn/present primitives were
        // wired into the supervisor by the daemon app.
        self.ctx.supervisor.lock().unwrap().open();
    }

    async fn get_snapshot(&self) -> String {
        self.ctx.engine.lock().await.snapshot_json()
    }

    async fn start_install(&self, spec: String) {
        let mut installs = self.ctx.installs.lock().await;
        if let Err(e) = installs.start(&spec) {
            log::warn!("Rejected install request: {e:#}");
        }
    }

    async fn get_installs(&self) -> String {
        self.ctx.installs.lock().await.running_json()
    }

    async fn quit(&self) {
        (self.ctx.on_quit)();
    }

    // ------------------------------------------------------------------
    // Signals pushed to the window (emitted by the daemon loop through the
    // generated Signals trait on the InterfaceRef).
    // ------------------------------------------------------------------

    #[zbus(signal)]
    async fn snapshot_changed(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
        json: String,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn error(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
        msg: String,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn output(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
        text: String,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn open_use_existing(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn present_window(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn install_progress(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
        key: String,
        text: String,
    ) -> zbus::Result<()>;

    #[zbus(signal)]
    async fn install_finished(
        signal_emitter: &zbus::object_server::SignalEmitter<'_>,
        key: String,
        ok: bool,
        message: String,
    ) -> zbus::Result<()>;
}
